### PR TITLE
Limit metrics collection of incoming requests

### PR DESCRIPTION
**Pull request by nosan**
see [gh-13913](https://github.com/spring-projects/spring-boot/issues/13913).

### CODE REVIEW

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/metrics/test/MetricsRun.java [hunk 1]
```diff
@@ -90,7 +91,7 @@ private MetricsRun() {
 	 * implementation.
 	 * @return the function to apply
 	 */
-	public static Function<ApplicationContextRunner, ApplicationContextRunner> simple() {
+	public static <T extends AbstractApplicationContextRunner> Function<T, T> simple() {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/metrics/test/MetricsRun.java]: This is not strictly related to the change. I am happy to revist the tests here but I'd like to do that in a separate PR. 

Could you please revert everything that's related to that?

(also if submit a new PR for this, there is a raw type use of `AbstractApplicationContextRunner`

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/web/servlet/WebMvcMetricsAutoConfigurationTests.java [hunk 2]
```diff
@@ -20,86 +20,117 @@
...
 
 /**
  * Tests for {@link WebMvcMetricsAutoConfiguration}.
  *
  * @author Andy Wilkinson
+ * @author Dmytro Nosan
  */
 public class WebMvcMetricsAutoConfigurationTests {
 
 	private WebApplicationContextRunner contextRunner = new WebApplicationContextRunner()
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/web/servlet/WebMvcMetricsAutoConfigurationTests.java]: Actually I prefer the former. I'd rather keep the auto-configurations explicit and only add what's strictly necessary for the task at hand (anyway, to be discussed in a separate PR anyway).

File: spring-boot-project/spring-boot-docs/src/main/asciidoc/appendix-application-properties.adoc [hunk 3]
```diff
@@ -1489,6 +1489,8 @@ content into your application. Rather, pick only the properties that you need.
 	management.metrics.web.client.requests-metric-name=http.client.requests # Name of the metric for sent requests.
 	management.metrics.web.server.auto-time-requests=true # Whether requests handled by Spring MVC or WebFlux should be automatically timed.
 	management.metrics.web.server.requests-metric-name=http.server.requests # Name of the metric for received requests.
+	management.metrics.web.server.max-uri-tags=100 # Maximum number of unique URI tag values allowed. After the max number of tag values is reached, metrics with additional tag values are denied by filter.
+
```
Comment (snicoll) [inline spring-boot-project/spring-boot-docs/src/main/asciidoc/appendix-application-properties.adoc]: Extra line to be removed here please.

File: spring-boot-project/spring-boot-docs/src/main/asciidoc/appendix-application-properties.adoc [hunk 4]
```diff
@@ -1489,6 +1489,8 @@ content into your application. Rather, pick only the properties that you need.
 	management.metrics.web.client.requests-metric-name=http.client.requests # Name of the metric for sent requests.
 	management.metrics.web.server.auto-time-requests=true # Whether requests handled by Spring MVC or WebFlux should be automatically timed.
 	management.metrics.web.server.requests-metric-name=http.server.requests # Name of the metric for received requests.
+	management.metrics.web.server.max-uri-tags=100 # Maximum number of unique URI tag values allowed. After the max number of tag values is reached, metrics with additional tag values are denied by filter.
```
Comment (snicoll) [inline spring-boot-project/spring-boot-docs/src/main/asciidoc/appendix-application-properties.adoc]: Please order the key alphabetically.

**Comment by snicoll** [comment]
@nosan thanks for the feedback but I still see some unrelated changes in `WebMvcMetricsAutoConfigurationTests`. Can you please remove those? The test you've added there imports `WebMvcMetricsAutoConfiguration` again. 

If you rebase to a single commit, then a diff should show a focused minimal change for this test as well. Thanks a lot for your efforts!

**Comment by nosan** [comment]
@snicoll 
updated :)

**Comment by nosan** [comment]
@snicoll I see that `WebClientMetricsAutoConfiguration` doesn't have a `MeterFilter` for `uri` tags as well. Should I add it?

**Comment by snicoll** [comment]
@nosan ah ah that's awesome, I just saw that as well 5 min ago. I think so yeah, looks like an oversight. This should be done as a separate issue/PR though. Do you want to give that a try? (I am almost done with this one).

**Comment by nosan** [comment]
@snicoll I'll create a separate `PR`.

**Comment by snicoll** [comment]
@nosan I am confused, there is a test for this in `WebClientMetricsAutoConfigurationTests`

**Comment by nosan** [comment]
@snicoll RestTemplateMetricsAutoConfiguration
```
	@Bean
	@Order(0)
	public MeterFilter metricsHttpClientUriTagFilter(MetricsProperties properties) {
		String metricName = properties.getWeb().getClient().getRequestsMetricName();
		MeterFilter denyFilter = new MaximumUriTagsReachedMeterFilter(metricName);
		return MeterFilter.maximumAllowableTags(metricName, "uri",
				properties.getWeb().getClient().getMaxUriTags(), denyFilter);
	}
```

**Comment by snicoll** [comment]
I don't understand why you refer to it. Metrics support for `WebClient` was added in #12228.

**Comment by nosan** [comment]
I'm talking about 'uri' tag filter. If you remove `RestTemplateMetricsAutoConfiguration`  from `MetricsRun.simple()` in `WebClientMetricsAutoConfigurationTests` then `afterMaxUrisReachedFurtherUrisAreDenied` will fail.
